app/controllers/BalanceRequestResponseController.scala

```diff
@@ -76,29 +78,52 @@ class BalanceRequestResponseController @Inject() (
     implicit request =>
       withMetricsTimerResult(UpdateBalanceRequest) {
         requireMessageTypeHeader { messageType =>
-          cache
-            .updateBalance(recipient, messageType, request.body)
-            .flatTap(logServiceError("updating balance request", _))
-            .map {
-              case Right(_) =>
-                Ok
-              case Left(error @ BadRequestError(_)) =>
-                BadRequest(Json.toJson[BalanceRequestError](error))
-              case Left(error @ XmlValidationError(_, _)) =>
-                BadRequest(Json.toJson[BalanceRequestError](error))
-              case Left(error @ NotFoundError(_)) =>
-                NotFound(Json.toJson[BalanceRequestError](error))
-              case Left(_) =>
-                InternalServerError(Json.toJson(BalanceRequestError.internalServiceError()))
-            }
-            .recoverWith { case NonFatal(e) =>
-              logger.error(e)("Unhandled exception thrown").map { _ =>
-                val error     = InternalServiceError.causedBy(e)
-                val errorJson = Json.toJson[BalanceRequestError](error)
-                InternalServerError(errorJson)
-              }
-            }
+          for {
+            result <- updateBalanceRequestIO(recipient, messageType)
+            _      <- recordRequestLength(request)
+          } yield result
         }
       }
   }
+
+  private def updateBalanceRequestIO(
+    recipient: MessageIdentifier,
+    messageType: MessageType
+  )(implicit request: Request[String], hc: HeaderCarrier): IO[Result] =
+    cache
+      .updateBalance(recipient, messageType, request.body)
+      .flatTap(logServiceError("updating balance request", _))
+      .map {
+        case Right(_) =>
+          Ok
+        case Left(error @ BadRequestError(_)) =>
+          BadRequest(Json.toJson[BalanceRequestError](error))
+        case Left(error @ XmlValidationError(_, _)) =>
+          BadRequest(Json.toJson[BalanceRequestError](error))
+        case Left(error @ NotFoundError(_)) =>
+          NotFound(Json.toJson[BalanceRequestError](error))
+        case Left(_) =>
+          InternalServerError(Json.toJson(BalanceRequestError.internalServiceError()))
+      }
+      .recoverWith { case NonFatal(e) =>
+        logger.error(e)("Unhandled exception thrown").map { _ =>
+          val error     = InternalServiceError.causedBy(e)
+          val errorJson = Json.toJson[BalanceRequestError](error)
+          InternalServerError(errorJson)
+        }
+      }
+
+  // TODO: do we want this to be a histogram?
```

Comment:
I think a histogram is fine - certainly that is the approach we have used in CTC Traders. Unfortunately the metrics system (Dropwizard Metrics -> Clickhouse Graphite API) cannot handle dimensional metrics so it's not a true histogram - it only produces pre-bucketed data at p50, p75, p90, p95, p99 etc, but we can show the pre-bucketed data over time.